### PR TITLE
If a Statement is a Static Invoke, Check Method Reference Declaring Class

### PR DESCRIPTION
**Pull request by JSAbrahams**
In the InsufficientLibraryUsageFilterRule filter. This fix ensures that we also include patterns which only have static calls the library, which would otherwise have been filtered away.

Furthermore:
* When accumulating visits, return true iff either is true

### CODE REVIEW

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 1]
```diff
@@ -34,21 +35,23 @@ class InsufficientLibraryUsageFilter(libraryProject: JavaProject, private val mi
...
      * @param value a [Value]
      * @return true iff the type of [value] is for a library class
      */
-    override fun applyDefault(value: Value) = isLibraryClass(value.type.toString())
+    override fun applyDefault(value: Value) = isLibraryClass(value.type.toString()) ||
+        (value is JStaticInvokeExpr) && isLibraryClass(value.methodRef.declaringClass().toString())
 
     /**
-     * Returns true iff [result1] and [result2] are both true.
+     * Returns true iff [result1] or [result2] are true.
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: Should be `is true`

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 2]
```diff
@@ -34,21 +35,23 @@ class InsufficientLibraryUsageFilter(libraryProject: JavaProject, private val mi
...
-     * Returns true iff the type of [value] is for a library class.
+     * Returns true iff the type of [value] is for a library class or [JStaticInvokeExpr] declaring class is for a
+     * library.
      *
      * @param value a [Value]
      * @return true iff the type of [value] is for a library class
      */
-    override fun applyDefault(value: Value) = isLibraryClass(value.type.toString())
+    override fun applyDefault(value: Value) = isLibraryClass(value.type.toString()) ||
+        (value is JStaticInvokeExpr) && isLibraryClass(value.methodRef.declaringClass().toString())
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: Instead of changing this method, you could add another method specifically for `StaticInvokeExpr`s:

```kotlin
override fun apply(value: StaticInvokeExpr) = isLibraryClass(value.methodRef.declaringClass().toString())
```

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 3]
```diff
@@ -34,21 +36,30 @@ class InsufficientLibraryUsageFilter(libraryProject: JavaProject, private val mi
  */
 class LibraryUsageVisitor(private val libraryProject: JavaProject) : JimpleValueVisitor<Boolean>() {
     /**
-     * Returns true iff the type of [value] is for a library class.
+     * Returns true iff the type of [value] is for a library class or [JStaticInvokeExpr] declaring class is for a
+     * library.
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: Looks like you forgot to change this documentation back to the original :/

File: modules/mining-pipeline/jimple-pattern-filter/src/test/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/libraryusagefilters/ValueFilterTest.kt [hunk 4]
```diff
@@ -140,7 +139,8 @@ internal object ValueFilterTest : Spek({
         }
 
         it("does not recognize unknown new expressions") {
-            assertThatItDoesNotRecognize(mock<AnyNewExpr>())
+            assertThatItDoesNotRecognize(
+                mock<AnyNewExpr>())
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/test/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/libraryusagefilters/ValueFilterTest.kt]: This class has quite a few formatting errors as a result of the refactoring.

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/libraryusagefilters/BranchStatementFilter.kt [hunk 5]
```diff
@@ -28,7 +28,7 @@ internal class BranchStatementFilter(project: JavaProject) : Filter {
             changed = false
 
             body.units.snapshotIterator().asSequence()
-                .filter(::isBranchStatement)
+                .filter(Companion::isBranchStatement)
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/libraryusagefilters/BranchStatementFilter.kt]: Why was this added?

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 6]
```diff
@@ -2,59 +2,29 @@ package org.cafejojo.schaapi.miningpipeline.patternfilter.jimple
...
  * @property minUseCount the minimum number of usages per pattern to be qualified as a "sufficient user"
  */
-class InsufficientLibraryUsageFilter(libraryProject: JavaProject, private val minUseCount: Int) :
-    PatternFilterRule<JimpleNode> {
-    private val libraryUsageVisitor = LibraryUsageVisitor(libraryProject)
-
+class InsufficientLibraryUsageFilterRule(
+    private val libraryProject: JavaProject,
+    private val minUseCount: Int
+) : PatternFilterRule<JimpleNode> {
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: This one is formatted different from how it was before. Was that intentional?

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt [hunk 7]
```diff
@@ -2,59 +2,29 @@ package org.cafejojo.schaapi.miningpipeline.patternfilter.jimple
...
-    /**
-     * Returns true iff [result1] and [result2] are both true.
-     *
-     * @param result1 a [Boolean]
-     * @param result2 a [Boolean]
-     * @return true iff [result1] and [result2] are both true
-     */
-    override fun accumulate(result1: Boolean, result2: Boolean) = result1 && result2
+    override fun retain(pattern: Pattern<JimpleNode>): Boolean {
+        val statementFilter = StatementFilter(libraryProject)
```
Comment (FWDekker) [inline modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: Since the filter rule is created per library project, this filter can be made into a field like in the other filter rules.

File: modules/mining-pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/miningpipeline/usagegraphgenerator/jimple/Helpers.kt [hunk 8]
```diff
@@ -1,8 +1,20 @@
...
+import com.nhaarman.mockito_kotlin.doReturn
+import com.nhaarman.mockito_kotlin.mock
 import org.cafejojo.schaapi.models.project.JavaProject
+import soot.Modifier
+import soot.SootClass
+import soot.SootMethod
+import soot.jimple.StaticInvokeExpr
 import java.io.File
 
+internal const val NON_LIBRARY_CLASS = "java.lang.String"
```
[the remark this answers is not in the document]
Comment (FWDekker) [reply modules/mining-pipeline/jimple-library-usage-graph-generator/src/test/kotlin/org/cafejojo/schaapi/miningpipeline/usagegraphgenerator/jimple/Helpers.kt]: Though I can certainly live with it remaining in the codebase. Even though it's unused, it feels somewhat inconsistent with the other helpers, and it may make it easier to write tests in this package.

File: modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt (hunk 6 above)
Comment (casperboone) [reply modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: Yes, I didn't like the original formatting very much

Comment (FWDekker) [reply modules/mining-pipeline/jimple-pattern-filter/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/patternfilter/jimple/InsufficientLibraryUsageFilterRule.kt]: I reverted it to the old style in 90024ca under the assumption that it was accidental. I have reverted the revert.
